include/rocksdb/write_batch.h

```diff
@@ -333,18 +336,43 @@ class WriteBatch : public WriteBatchBase {
   // Returns true if MarkRollback will be called during Iterate
   bool HasRollback() const;
 
-  // Assign timestamp to write batch.
-  // This requires that all keys (possibly from multiple column families) in
-  // the write batch have timestamps of the same format.
-  Status AssignTimestamp(const Slice& ts);
+  struct TimestampChecker final {
+    Status operator()(uint32_t /*cf*/, size_t& /*ts_sz*/) const {
+      return Status::OK();
+    }
+  };
 
+  // Experimental.
+  // Assign timestamp to write batch.
```

Comment:
Do you mean overwrite the timestamp of all existing entries in the write batch?

Reply:
Not all entries. If a key is for a column family that disables timestamp, it will skip this key. Since all keys in the write batch, if enable timestamp, share the same timestamp, I think it's OK to say "assign a timestamp to the write batch".

Reply:
> Since all keys in the write batch, if enable timestamp, share the same timestamp

Where does this come from? The API documentation for WriteBatch::Put suggest you can assign a timestamp per key.

Regarding that: the API doc for `Status Put(ColumnFamilyHandle* column_family, const Slice& key, const Slice& value)` suggests it supports timestamp, which there is no test for that I can find.

From API:
> as long as `key` points to a contiguous buffer with timestamp appended after user key.

The way this is phrased suggests to me that space for the timestamp has to be part of the buffer pointed to by the Slice, but beyond the bounds (size) of the Slice. (To match DB::Put, key Slice size does not include timestamp.)

Also,
> as long as the timestamp is the last Slice in (SliceParts) `key`

I don't see such a limitation in the implementation, and it seems to break the intended flexibility / encapsulation provided by SliceParts: the user of SliceParts should treat it logically as one contiguous data buffer. ("A set of Slices that are virtually concatenated together.")

Reply:
We are talking about `AssignTimestamp(const Slice& ts)` not `Put`. The result of this API is that all keys, if enable timestamp, will have the same timestamp.

Reply:
> The result of this API is that all keys, if enable timestamp, will have the same timestamp.

What about keys added after `AssignTimestamp`? Your description doesn't make it clear. If the data model were more clear to the API user, this might be obvious, but the data model with timestamps is not clear from the API comments. The name "AssignTimestamp" alone (not to mention "Assign timestamp to write batch") suggests a data model in which a timestamp is tracked separately from individual entries to apply to them at Write time. I had to look at the implementations to understand the contract for AssignTimestamp(s) (and Put).

Reply:
The keys that do not exist in the WB when calling the `AssignTimestamp` API will obviously not have their timestamp set to the value. We can make it clear in the API comments.
Since the `WriteBatch::Handler` is public, user can always define their own handler, call `Iterate()` and then insert new keys after that. Therefore, your concern applies to the entire `WriteBatch::Handler`, and I do not have a good way of preventing that.